### ISSUE TITLE
podman remote flag/overlay check (regression?)

### ISSUE THREAD

**Issue by 42wim** [containers/podman#7487]
/kind bug

**Description**

Running podman --remote in a container doesn't work anymore in v2.0.5 (worked in v2.0.1) (not tested on versions between v2.0.1 and v2.0.5 yet)

**Steps to reproduce the issue:**

1. run a podman remote service (podman system service --timeout=0)

2. run a container which bind mounts the socket (/run/podman/podman.sock)

3. run podman <any command> --remote (eg podman images --remote)

**Describe the results you received:**

```
podman images --remote
Error: unknown flag: --remote
```

```
podman --remote images
Error: 'overlay' is not supported over overlayfs, a mount_program is required: backing file system is unsupported for this graph driver
```

**Describe the results you expected:**

this is using v2.0.1 in the container

```
bash-4.4# podman images --remote
REPOSITORY                                              TAG     IMAGE ID      CREATED       SIZE
registry.centos.org/centos                              8       0cfc5d6fad87  3 months ago  228 MB
registry.centos.org/centos                              7       8114e7c1868b  3 months ago  210 MB
docker.io/library/ubuntu                                latest  1d622ef86b13  4 months ago  76.3 MB
```

**Additional information you deem important (e.g. issue happens only occasionally):**

This worked fine in podman v2.0.1
I've upgraded to v2.0.5 and it stopped working

**Output of `podman version`:**

```
podman version
Error: 'overlay' is not supported over overlayfs, a mount_program is required: backing file system is unsupported for this graph driver
```

**Output of `podman info --debug`:**

```
podman info --debug
Error: 'overlay' is not supported over overlayfs, a mount_program is required: backing file system is unsupported for this graph driver
```

**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**

```
rpm -q podman
podman-2.0.5-1.el8.x86_64
```

**Have you tested with the latest version of Podman and have you checked the Podman Troubleshooting Guide?**

Yes, latest (released) version of podman, no not checked podman troubleshooting guide.

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Running podman in a container talking to a remote podman on the same host using bind mount podman socket.

**Comment by 42wim** [comment]
Seems older rpm's aren't kept on https://download.opensuse.org/repositories/devel:/kubic:/libcontainers:/stable/CentOS_7/x86_64/ so I can't easily try an older version.

But my guess is https://github.com/containers/podman/pull/6716 has something to do with this issue

**Comment by Luap99** [comment]
I think this is by design. `podman subcommand --remote` does not make much sense as `--remote` is only applicable to podman and not the subcommand. That fact that this worked before introduced some bugs and that's why it got changed with #6716.


```
Error: 'overlay' is not supported over overlayfs, a mount_program is required: backing file system is unsupported for this graph driver
```
That could be a invalid config. Looks like you are running as root, can you share the output of `/etc/containers/storage.conf`?

**Comment by 42wim** [comment]
@luap99 The same config and command works fine on podman v2.0.1
The /etc/containers/storage.conf is the default config from the containers-common-1.1.1-1.el8.x86_64 rpm
Everything is default, no config has been changed.

**Comment by Luap99** [comment]
Can you try `podman --storage-driver=vfs info` please.

**Comment by 42wim** [comment]
```
podman --storage-driver=vfs info
Error: database storage graph driver "overlay" does not match our storage graph driver "vfs": database configuration mismatch
```

**Comment by Luap99** [comment]
Can you create a new container and remove the storage.conf before you run any podman commands.

**Comment by 42wim** [comment]
This can be very easily reproduced, you don't even need a remote podman running as it doesn't look at that yet because of the overlay issues

```
# podman run --rm -ti centos:8
```

```
[root@815647e90058 /]# dnf -y module disable container-tools && dnf -y install 'dnf-command(copr)' && dnf -y copr enable rhcontainerbot/container-selinux && curl -L -o /etc/yum.repos.d/devel:kubic:libcontainers:stable.repo https://download.opensuse.org/repositories/devel:/kubic:/libcontainers:/stable/CentOS_8/devel:kubic:libcontainers:stable.repo
[root@815647e90058 /]# yum -y install podman
[root@815647e90058 /]# podman --remote ps
Error: 'overlay' is not supported over overlayfs, a mount_program is required: backing file system is unsupported for this graph driver
[root@815647e90058 /]# podman --storage-driver=vfs info
Error: database storage graph driver "overlay" does not match our storage graph driver "vfs": database configuration mismatch
```

**Comment by 42wim** [comment]
when removing the storage.conf in a new container before any commands:

```
[root@6e2eb5a34cf4 /]# podman --storage-driver=vfs info
host:
  arch: amd64
  buildahVersion: 1.15.1
  cgroupVersion: v1
  conmon:
    package: conmon-2.0.20-1.el8.x86_64
    path: /usr/bin/conmon
    version: 'conmon version 2.0.20, commit: 064c4ad2d7293c326be702ec8fcae8a3741aa6ee'
  cpus: 24
  distribution:
    distribution: '"centos"'
    version: "8"
  eventLogger: file
  hostname: 6e2eb5a34cf4
  idMappings:
    gidmap: null
    uidmap: null
  kernel: 4.18.0-193.6.3.el8_2.x86_64
  linkmode: dynamic
  memFree: 62486818816
  memTotal: 67251437568
  ociRuntime:
    name: runc
    package: runc-1.0.0-145.rc91.git24a3cf8.el8.x86_64
    path: /usr/bin/runc
    version: 'runc version spec: 1.0.2-dev'
  os: linux
  remoteSocket:
    path: /run/podman/podman.sock
  rootless: false
  slirp4netns:
    executable: ""
    package: ""
    version: ""
  swapFree: 2147479552
  swapTotal: 2147479552
  uptime: 1582h 27m 44.2s (Approximately 65.92 days)
registries:
  search:
  - registry.fedoraproject.org
  - registry.access.redhat.com
  - registry.centos.org
  - docker.io
store:
  configFile: /etc/containers/storage.conf
  containerStore:
    number: 0
    paused: 0
    running: 0
    stopped: 0
  graphDriverName: vfs
  graphOptions: {}
  graphRoot: /var/lib/containers/storage
  graphStatus: {}
  imageStore:
    number: 0
  runRoot: /var/run/containers/storage
  volumePath: /var/lib/containers/storage/volumes
version:
  APIVersion: 1
  Built: 0
  BuiltTime: Thu Jan  1 00:00:00 1970
  GitCommit: ""
  GoVersion: go1.13.4
  OsArch: linux/amd64
  Version: 2.0.5
```

**Comment by 42wim** [comment]
Did some tests now with bind mounting the socket again.
(`podman run -v /run/podman:/run/podman  --rm -ti centos:8`)

Now it doesn't give a VFS error anymore, but another error and not the correct output (it should show the centos:8 image)

```
[root@ff68a17c8580 /]# rm /etc/containers/storage.conf -f
[root@ff68a17c8580 /]# podman --remote images
ERRO[0000] Failed to built-in GetDriver graph btrfs /var/lib/containers/storage
REPOSITORY  TAG     IMAGE ID  CREATED  SIZE
```

**Comment by mheon** [comment]
Can you add `--log-level=debug` to one of these commands, so we can verify it is indeed ignoring `--remote`?

@jwhonce PTAL

**Comment by 42wim** [comment]
```
[root@cbdd165e08a0 /]# podman --log-level=debug  --remote images
INFO[0000] podman filtering at log level debug
DEBU[0000] Called images.PersistentPreRunE(podman --log-level=debug --remote images)
DEBU[0000] Ignoring libpod.conf EventsLogger setting "/etc/containers/containers.conf". Use "journald" if you want to change this setting and remove libpod.conf files.
DEBU[0000] Reading configuration file "/usr/share/containers/containers.conf"
DEBU[0000] Merged system config "/usr/share/containers/containers.conf": &{{[] [] containers-default-0.14.9 [] host enabled [CAP_AUDIT_WRITE CAP_CHOWN CAP_DAC_OVERRIDE CAP_FOWNER CAP_FSETID CAP_KILL CAP_MKNOD CAP_NET_BIND_SERVICE CAP_NET_RAW CAP_SETFCAP CAP_SETGID CAP_SETPCAP CAP_SETUID CAP_SYS_CHROOT] [] [nproc=4194304:4194304]  [] [] [] false [PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin TERM=xterm] false false false  private k8s-file -1 bridge false 2048 private /usr/share/containers/seccomp.json 65536k private host 65536} {false systemd [PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin] [/usr/libexec/podman/conmon /usr/local/libexec/podman/conmon /usr/local/lib/podman/conmon /usr/bin/conmon /usr/sbin/conmon /usr/local/bin/conmon /usr/local/sbin/conmon /run/current-system/sw/bin/conmon] ctrl-p,ctrl-q true /var/run/libpod/events/events.log file [/usr/share/containers/oci/hooks.d] docker:// /pause k8s.gcr.io/pause:3.2 /usr/libexec/podman/catatonit shm   false 2048 runc map[crun:[/usr/bin/crun /usr/sbin/crun /usr/local/bin/crun /usr/local/sbin/crun /sbin/crun /bin/crun /run/current-system/sw/bin/crun] kata:[/usr/bin/kata-runtime /usr/sbin/kata-runtime /usr/local/bin/kata-runtime /usr/local/sbin/kata-runtime /sbin/kata-runtime /bin/kata-runtime /usr/bin/kata-qemu /usr/bin/kata-fc] runc:[/usr/bin/runc /usr/sbin/runc /usr/local/bin/runc /usr/local/sbin/runc /sbin/runc /bin/runc /usr/lib/cri-o-runc/sbin/runc /run/current-system/sw/bin/runc]] missing false    map[] [] [crun runc] [crun] [kata kata-runtime kata-qemu kata-fc] {false false false false false false} /etc/containers/policy.json false 3 /var/lib/containers/storage/libpod 10 /var/run/libpod /var/lib/containers/storage/volumes} {[/usr/libexec/cni /usr/lib/cni /usr/local/lib/cni /opt/cni/bin] podman /etc/cni/net.d/}}
DEBU[0000] Using conmon: "/usr/bin/conmon"
DEBU[0000] Initializing boltdb state at /var/lib/containers/storage/libpod/bolt_state.db
DEBU[0000] Using graph driver vfs
DEBU[0000] Using graph root /var/lib/containers/storage
DEBU[0000] Using run root /var/run/containers/storage
DEBU[0000] Using static dir /var/lib/containers/storage/libpod
DEBU[0000] Using tmp dir /var/run/libpod
DEBU[0000] Using volume path /var/lib/containers/storage/volumes
DEBU[0000] Set libpod namespace to ""
DEBU[0000] [graphdriver] trying provided driver "vfs"
DEBU[0000] Initializing event backend file
DEBU[0000] using runtime "/usr/bin/runc"
WARN[0000] Error initializing configured OCI runtime crun: no valid executable found for OCI runtime crun: invalid argument
WARN[0000] Error initializing configured OCI runtime kata: no valid executable found for OCI runtime kata: invalid argument
INFO[0000] Found CNI network podman (type=bridge) at /etc/cni/net.d/87-podman-bridge.conflist
WARN[0000] Default CNI network name podman is unchangeable
INFO[0000] Setting parallel job count to 73
REPOSITORY  TAG     IMAGE ID  CREATED  SIZE
DEBU[0000] Called images.PersistentPostRunE(podman --log-level=debug --remote images)
```

**Comment by mheon** [comment]
I see libpod debug info in there, so it looks like `--remote` is indeed being ignored.

**Comment by Luap99** [comment]
I found it. https://github.com/containers/podman/pull/7042 is not backported to 2.0.

**Comment by Luap99** [comment]
I opened #7491 to backport this.

**Comment by afbjorklund** [comment]
Same here, using `podman-remote` as the fallback since `podman --remote` is so "unstable"

```
 podman2 --remote version
Version:      2.0.5
API Version:  1
Go Version:   go1.13.8
Git Commit:   776abc52106ec7652ced6dbc0869020123ed393d
Built:        Sat Aug 29 16:38:22 2020
OS/Arch:      linux/amd64
```

```
 podman2-remote version
Client:
Version:      2.0.5
API Version:  1
Go Version:   go1.13.8
Git Commit:   776abc52106ec7652ced6dbc0869020123ed393d
Built:        Sat Aug 29 16:38:29 2020
OS/Arch:      linux/amd64

Server:
Version:      2.0.5
API Version:  0
Go Version:   go1.14.6
Built:        Thu Jan  1 01:00:00 1970
OS/Arch:      linux/amd64
```

But there are no guarantees, since podman-remote also runs locally "sometimes". https://github.com/containers/podman/issues/6598#issuecomment-651800762

It would be nice if there had been a simple regression test, for each release (especially v2).
Like checking the basic podman-remote version/run/build/images commands or something ?

But as it is now, recommendation is to stay on 1.9 until v2 becomes "stable" - maybe 2.1 ?

**Comment by mheon** [comment]
Going to close this, as the relevant patches are now included in 2.0.6
